[GEP-34] Opentelemetry Operator and Collectors

nickytd opened the pull request:
**How to categorize this PR?**
/area logging
/area monitoring
/kind enhancement

**What this PR does / why we need it**:
This PR introduces GEP-34 bringing OpentTelemetry support in Gardener, part of Observability 2.0 initiative.

**Release note**:

```other developer
GEP-34 Introducing OpenTelemetry Operator and Collectors in Shoot Control Planes
```

nickytd committed:
gep-34: initial gep proposal

Kostov6 reviewed (commented):
Thanks for the GEP! I was wondering if the migration will unavoidably lead to some loss of logs. If yes do you have some sort of a rough estimation how much logs will be lost?

ScheererJ commented on docs/proposals/34-observability2.0-opentelemtry-operator-and-collectors.md:
```diff
@@ -0,0 +1,156 @@
+
+- ""
+
+---
```
Please use the default header. The existing one looks weird compared to the other GEPs.

```suggestion
---
title: Introducing OpenTelemetry Operator and Collectors in Shoot Control Planes
gep-number: 34
creation-date: 2025-04-12
status: implementable
authors:
- "@nickytd"
- "@rrhubenov"
reviewers:
- 
---

# GEP-[34]: Introducing OpenTelemetry Operator and Collectors in Shoot Control Planes
```

ScheererJ commented on docs/proposals/34-observability2.0-opentelemtry-operator-and-collectors.md:
```diff
@@ -0,0 +1,156 @@
+
+## Motivation
+
+Gardener's current observability stack, while improved by the operator-based approach in *GEP-19*, still relies on vendor specific format and protocols for collecting, processing and storing observability signals. This brings challenges in extending both components and scenarios used to process these signals and enforces lock-in integration scenarios with external consumers. By introducing the Op ...
```
```suggestion
Gardener's current observability stack, while improved by the operator-based approach in [GEP-19](https://github.com/gardener/gardener/blob/master/docs/proposals/19-migrating-observability-stack-to-operators.md), still relies on vendor specific format and protocols for collecting, processing and storing observability signals. This brings challenges in extending both components and scenarios used to process these signals and enforces lock-in integration scenarios with external consumers. By introducing the OpenTelemetry operator and collectors, this proposal addresses these challenges, aligning Gardener with industry trends and preparing it for future observability enhancements, such as unified visualization and tracing support.
```

ScheererJ commented on docs/proposals/34-observability2.0-opentelemtry-operator-and-collectors.md:
```diff
@@ -0,0 +1,156 @@
+- **Deploy the OpenTelemetry Operator on Seed Clusters:** Install and configure the OpenTelemetry operator to manage otel-collector instances across shoot control plane namespaces.
+- **Deploy Gardener OpenTelemetry Collector Distribution in Shoot Control Planes:** Provision a dedicated Gardener OpenTelemetry collector distribution instance via the OpenTelemetry Operator in each shoot control plane namespace to handle observability data.
+- **Transport Logs signals through the OpenTelemetry Collector Distribution:** Configure fluent-bits to send logs to the OTel Collectors receivers and wire the otel collector exporters to the current logging backend.
+- **Replace valitail system service with otel-collectors on shoot nodes:** Replace current valitail system service on shoot nodes with otel-collector, keeping parity in the collected type of logs.
```
You use various variations of similar terms/casing, e.g. OpenTelemetry Operator, OpenTelemetry operator, otel-collector, OpenTelemetry Collector, OpenTelemetry collector, OTel Collectors.
Could you please stick to less variations and be consistent with the casing? Unless of course those mean very different things, which is not obvious for me as a reader.

nickytd replied:
thanks for the proposal, will adjust the casing.

ScheererJ commented on docs/proposals/34-observability2.0-opentelemtry-operator-and-collectors.md:
```diff
@@ -0,0 +1,156 @@
+- **Deploy the OpenTelemetry Operator on Seed Clusters:** Install and configure the OpenTelemetry operator to manage otel-collector instances across shoot control plane namespaces.
+- **Deploy Gardener OpenTelemetry Collector Distribution in Shoot Control Planes:** Provision a dedicated Gardener OpenTelemetry collector distribution instance via the OpenTelemetry Operator in each shoot control plane namespace to handle observability data.
+- **Transport Logs signals through the OpenTelemetry Collector Distribution:** Configure fluent-bits to send logs to the OTel Collectors receivers and wire the otel collector exporters to the current logging backend.
+- **Replace valitail system service with otel-collectors on shoot nodes:** Replace current valitail system service on shoot nodes with otel-collector, keeping parity in the collected type of logs.
```
Maybe I misread the goals, but I fail to understand what the boundary conditions of this proposal are. What do you want to achieve? Should all observability data go through open telemetry collectors? Do we only care about the data coming from one shoot cluster? Do we handle both signals from the seed (control plane) as well as the shoot (data plane)? Are all signals in scope, i.e. do we want to adapt logging and metrics or only one of them?
Please clarify.

nickytd replied:
The evolution of the topic depends on the concrete scenarios. For metrics for example I plan to bring a prposal for the  hackaton in June hack-the-garden-25 event https://gardener-cloud.slack.com/archives/C0531FVMZFU to use the otel-collector instance on the shoot nodes to also scrape the exportes and push the metrics in the otel-collector in the control-plane namespace and from there to the prometheus 3 instances. The prometheus 3 supports standard otel ingegstion. https://prometheus.io/blog/2024/03/14/commitment-to-opentelemetry/

Short terms goals: is to transition the logs. Shoot metrics are potential candidate that we shall explore next also with the collaboratin with the monitoring team. For this early stage no other changes for metrics are planned.

Long term goal: is to potentially transition all observbaility signals but that will be motivated from the potential scenarios. That is already explained.

ScheererJ commented on docs/proposals/34-observability2.0-opentelemtry-operator-and-collectors.md:
```diff
@@ -0,0 +1,156 @@
+
+- **Immediate Replacement of Existing Tools:** This proposal does not aim to immediately decommission tools like Vali, Fluent-bit, or Prometheus; the migration will be phased.
+- **Unified Visualization Implementation:** Developing a single UI for logs, metrics, and traces is out of scope, as no mature open-source solution currently exists (per *Observability 2.0*).
+- **Full Tracing Adoption:** While tracing support is enabled, widespread instrumentation of Gardener components for distributed traces is a future step.
```
Is tracing at all in focus or is this GEP just a precondition to adopt tracing more easily at a later point in time?

nickytd replied:
That has been clarified already. This GEP does not address distributed traces in gardener per se. But is a precondition as you correctly pointed for some time in the futre.

ScheererJ commented on docs/proposals/34-observability2.0-opentelemtry-operator-and-collectors.md:
```diff
@@ -0,0 +1,156 @@
+
+## Proposal
+
+This section outlines the implementation plan to introduce the OpenTelemetry operator and collectors into Gardener's shoot control plane namespaces and in shoot nodes as a systemd service, aligning with Step 2 of the *Observability 2.0* roadmap.
```
Please add a link.

nickytd replied:
What link?

vpnachev replied:
I guess https://github.com/gardener/logging/blob/master/docs/observability-2.0/Observability%202.0.md ?

ScheererJ commented on docs/proposals/34-observability2.0-opentelemtry-operator-and-collectors.md:
```diff
@@ -0,0 +1,156 @@
+### Non-Goals
+
+- **Immediate Replacement of Existing Tools:** This proposal does not aim to immediately decommission tools like Vali, Fluent-bit, or Prometheus; the migration will be phased.
+- **Unified Visualization Implementation:** Developing a single UI for logs, metrics, and traces is out of scope, as no mature open-source solution currently exists (per *Observability 2.0*).
```
Please add a link.

nickytd replied:
Sure but what link?

ScheererJ replied:
Probably just https://github.com/gardener/logging/blob/master/docs/observability-2.0/Observability%202.0.md unless you have something else in mind.
`Observability 2.0` by itself is not very helpful for people not in the details. Therefore, pointing to the overall story with a link helps a lot.

ScheererJ commented on docs/proposals/34-observability2.0-opentelemtry-operator-and-collectors.md:
```diff
@@ -0,0 +1,156 @@
+
+- **Objective:** Establish centralized management of OpenTelemetry collectors across all shoot control planes.
+- **Details:** Deploy the OpenTelemetry operator (e.g., from the [OpenTelemetry Operator for Kubernetes](https://github.com/open-telemetry/opentelemetry-operator)) in the `garden` namespace of each seed cluster using `ManagedResources`, similar to the Prometheus operator deployment in *GEP-19*.
+- **Configuration:** The operator will be deployed following the `botanist` workflow in the same way as the current `prometheus` and `fluent-bit` operators. This meaning that during the `Seed` reconciliation flow, the OpenTelemetryOperator will be deployed into the `garden` namespace.
```
```suggestion
- **Configuration:** The operator will be deployed following the `botanist` workflow in the same way as the current `prometheus` and `fluent-bit` operators. This means that during the `Seed` reconciliation flow the OpenTelemetryOperator will be deployed into the `garden` namespace.
```

ScheererJ commented on docs/proposals/34-observability2.0-opentelemtry-operator-and-collectors.md:
```diff
@@ -0,0 +1,156 @@
+
+- **Objective:** Provide each shoot control plane with a dedicated collector to process its observability data.
+- **Details:** For each shoot, deploy an OpenTelemetry collector instance in its control plane namespace (e.g., `shoot--project--name`). The collector will run as a `Deployment` managed by the OpenTelemetry operator.
+- Configuration: During the `Shoot` reconciliation flow, an [`OpenTelemetryCollector` Custom Resource](https://github.com/open-telemetry/opentelemetry-operator/blob/main/apis/v1beta1/opentelemetrycollector_types.go) will be created in the `Shoot` namespace. This Custom Resource will trigger the Operator to create an OpenTelemetry Collector Deployment & Ingress inside the namespace.
```
```suggestion
- **Configuration**: During the `Shoot` reconciliation flow, an [`OpenTelemetryCollector` Custom Resource](https://github.com/open-telemetry/opentelemetry-operator/blob/main/apis/v1beta1/opentelemetrycollector_types.go) will be created in the `Shoot` namespace. This Custom Resource will trigger the Operator to create an OpenTelemetry Collector Deployment & Ingress inside the namespace.
```

ScheererJ commented on docs/proposals/34-observability2.0-opentelemtry-operator-and-collectors.md:
```diff
@@ -0,0 +1,156 @@
+
+- **Objective:** Provide each shoot control plane with a dedicated collector to process its observability data.
+- **Details:** For each shoot, deploy an OpenTelemetry collector instance in its control plane namespace (e.g., `shoot--project--name`). The collector will run as a `Deployment` managed by the OpenTelemetry operator.
+- Configuration: During the `Shoot` reconciliation flow, an [`OpenTelemetryCollector` Custom Resource](https://github.com/open-telemetry/opentelemetry-operator/blob/main/apis/v1beta1/opentelemetrycollector_types.go) will be created in the `Shoot` namespace. This Custom Resource will trigger the Operator to create an OpenTelemetry Collector Deployment & Ingress inside the namespace.
```
Does it have to be an `Ingress` resource or is it also possible to use istio directly, either via istio APIs or Gateway API? The network flow through two reverse proxies is not perfect for `Ingress` flows.

nickytd replied:
I was thinking about that as well, if we can bring Gateway API into our landscapes that definelty is a viable transion step from the ingress. So we can start with the ingress because that is what we have today and transition to Gateway API when that become available. I would not bring a dependency to an istio specific API.

ScheererJ commented on docs/proposals/34-observability2.0-opentelemtry-operator-and-collectors.md:
```diff
@@ -0,0 +1,156 @@
+      endpoint: https://v-local--local.ingress.local.seed.local.gardener.cloud/vali/api/v1/push
+        headers:
+        # Authenticate against Vali using a Bearer Token
+        Authorization: "Bearer ${file:{{ .pathAuthToken }}}"
```
Indentation seems to be incorrect. Please fix.

nickytd replied:
Thanks will update the indentation.

ScheererJ commented on docs/proposals/34-observability2.0-opentelemtry-operator-and-collectors.md:
```diff
@@ -0,0 +1,156 @@
+      endpoint: https://v-local--local.ingress.local.seed.local.gardener.cloud/vali/api/v1/push
+        headers:
+        # Authenticate against Vali using a Bearer Token
+        Authorization: "Bearer ${file:{{ .pathAuthToken }}}"
```
How does the file include work? Is this a potential security vulnerability waiting to happen as any local file could be exfiltrated?

nickytd replied:
There is no difference than any other configuration or component that the gardener controllers manage.

rrhubenov replied:
During initialisation, the collector will replace the template with the contents of the file.

We looked into this more carefully and tried to determine whether this would be a security vulnerability. We came to the conclusion that since this file is hosted on the node itself, anyone who has access to it and can do whatever they want with it.
If an unwanted actor gets access to either file(the collector config or the file containing the token) then that would mean that they had already previously gotten access to the whole machine. The current `valitail` configuration does almost the same thing. Only difference is that `valitail` receives the path to the file with the auth token and reads it, rather than it being embedded in the config file itself.
There is a way for this to be done the same as it currently is in valitail, if that approach is determined to be more appropriate.

We can discuss this in more detail if you have concerns that we haven't been able to address.

rrhubenov replied:
Taking a step back and assessing the scenario, in the future implementation vali (or whatever log aggregator we'll use) won't be publicly accessible since the collector will theoretically act as the single point of access for logs (maybe all signals). 

That would mean that we can remove the ingress in front of Vali and remove the authentication as well.
The only authentication will be from the shoot nodes to the collector.

ScheererJ commented on docs/proposals/34-observability2.0-opentelemtry-operator-and-collectors.md:
```diff
@@ -0,0 +1,156 @@
+
+- **Objective:** Ensure a smooth transition by interfacing with the current observability stack.
+- **Details:**
+  - **Fluent-bit Integration:** Configure Fluent-bit to forward logs to the collector’s `loki` receiver, maintaining compatibility with the Vali-based setup from *GEP-19*.
```
Please add a link.

nickytd replied:
Link to what?

ScheererJ replied:
We usually link between GEPs in case they refer to each other. Hence, please add a link to GEP-19.

ScheererJ commented on docs/proposals/34-observability2.0-opentelemtry-operator-and-collectors.md:
```diff
@@ -0,0 +1,156 @@
+
+![Figure 1](assets/gep-34-otel-operator-architecture.png)
+
+### 3. Integrate with Existing Components
```
What about integration with prometheus?

nickytd replied:
Yes we are already thinking about potential scenarios. At this stage it is too early to make a concrete proposal. It was explained in other comment.

ScheererJ commented on docs/proposals/34-observability2.0-opentelemtry-operator-and-collectors.md:
```diff
@@ -0,0 +1,156 @@
+- **Details:**
+  - **Fluent-bit Integration:** Configure Fluent-bit to forward logs to the collector’s `loki` receiver, maintaining compatibility with the Vali-based setup from *GEP-19*.
+  - **Vali Integration:** Use the `loki` exporter to output the shoot logs to the existing Vali backend.
+  - **Valitail Replacement:** Replace the current `valitail` system service on shoot nodes with the OpenTelemetry collector, ensuring it collects the same types of logs as before. This will be done by deploying a systemd service for the collector on each shoot node, in the same way as the current `valitail` service.
```
How will this replacement work? Will the same systemd unit be used, i.e. will an update simply switch from one executable to the new one or shall both units work in parallel? Can you please share more details?

nickytd replied:
Current `valitail` service will be replaced with `otel-collector` systemd service. There is no need to have 2 systemd services running in parallel. An implementation is already in progress and we are working out the details. Will share those if still needed.

vpnachev replied:
Do you plan to change how the agents on the node connect and authenticate with the endpoint in the seeds?
Just for completeness, it would be nice to be mentioned if the current solution with the `kube-rbac-proxy` will be preserved or replaced with another one.

rfranzke replied:
@nickytd ?

nickytd replied:
We discussed this point in length. Here is the summary:
- Initial implementation in `gep-34` preserves the current authentication (service account) and authorization (rbac) mechanism implemented in the `valitail`. There will be no semantical change to what is employed as of today. Goal is to focus on enabling the e2e scenarios described in `gep-34`, while keeping the changes and the scope under control.

However the goal in the next iteration is to move towards mTLS, especially if [Gateway API](https://gateway-api.sigs.k8s.io/api-types/gateway/) kicks in and provides support for [TLSRoutes](https://gateway-api.sigs.k8s.io/reference/spec/#tlsroute) and backend TLS termination with [BackednTLSPolicy](https://gateway-api.sigs.k8s.io/api-types/backendtlspolicy/) That initiative is rather separate and solves the general question of introducing mTLS configurations for various cluster service consumptions. So yes we want to evolve the current auth & autz mechanism and potentially move towards mTLS after we implemented the `gep-34` proposals, Moving towards GatewayAPI will also open the opportunity for getting rid of the nginx ingress controller especially if we solve the prometheus federation constraint. So it is interesting problem to solve per se with a significant scope. But with `gep-34` we are getting closer and we are laying the foundations for evolving further in the right direction.

ScheererJ commented on docs/proposals/assets/gep-34-shoot-collector-config.yaml:
```diff
@@ -0,0 +1,133 @@
+          - key: _SYSTEMD_UNIT
+            value: kubelet.service
+          - key: _SYSTEMD_UNIT
+            value: docker.service
```
Probably not necessary anymore.

ScheererJ commented on docs/proposals/34-observability2.0-opentelemtry-operator-and-collectors.md:
```diff
@@ -0,0 +1,156 @@
+
+- **Description:** Deploy OpenTelemetry collectors in each shoot reconcile directly without using an operator.
+- **Pros:** Simplifies initial setup, avoids operator overhead.
+- **Cons:** Lacks scalability and management features (e.g., automatic upgrades, CR-based configuration), which are critical in Gardener’s multi-tenant environment.
```
Could you please explain in more detail the topic automatic upgrades in this context?

From my point of view, regardless if we use an operator or not there would be some version reference for the operator or the collector. This would need to be updated manually, e.g. via a renovate PR or a corresponding version upgrade. Once that change would make it into a gardener release the upgrades would be automatic for landscape operators/shoot owners.

What is the difference of both approaches in this regard? I do not see any.

nickytd replied:
You are correct. But here I refer to a decision that has been made in past to decouple the lifecycle management of the external (to gardener) components to community supoprted operators the same as `fluent-bit` or `prometheus`.

ScheererJ commented on docs/proposals/34-observability2.0-opentelemtry-operator-and-collectors.md:
```diff
@@ -0,0 +1,156 @@
+- **Description:** Deploy OpenTelemetry collectors in each shoot reconcile directly without using an operator.
+- **Pros:** Simplifies initial setup, avoids operator overhead.
+- **Cons:** Lacks scalability and management features (e.g., automatic upgrades, CR-based configuration), which are critical in Gardener’s multi-tenant environment.
+- **Reason Rejected:** The operator aligns with *GEP-19*’s precedent of using operators for observability components, ensuring consistency and scalability.
```
Frankly, I do not see the precedent of GEP-19 (link missing) as an important aspect.

An operator can make it easier to manage the resources, as was the case with GEP-19, but we do not use operators for all of our components.

nickytd replied:
I do find it important though, we do use operators to manage the lifecycle of non-gardener components. A reason is to remove these specifics from the gardener code base switching to community employed practices and more importantly to bring the declaritive aspects managing those componets and their specifics.

ScheererJ commented on docs/proposals/34-observability2.0-opentelemtry-operator-and-collectors.md:
```diff
@@ -0,0 +1,156 @@
+
+---
+
+This GEP establishes a critical milestone in Gardener’s journey toward *Observability 2.0*. By deploying the OpenTelemetry operator and collectors, Gardener gains a standardized, extensible observability framework that integrates with its existing stack while paving the way for future enhancements like tracing and OTLP-native backends.
```
While this GEP certainly is a stepping stone towards Observability 2.0, I would appreciate it if you would mention more explicitly what this is all about. In this GEP there is almost no mention of any signals but logging. This is fine, but then it should also be clear from reading the summary/motivation that this only handles logging.

Please adapt the document accordingly.

nickytd replied:
Thanks, the summary is updated as requested.

ScheererJ reviewed (changes requested):
Thanks for providing a glimpse into what is planned for observability 2.0.

Please take a minute to clarify the scope of the GEP. Title, summary and motivation made me think that this has far more reach than it actually intends to implement according to the proposal section.

nickytd committed:
Update docs/proposals/34-observability2.0-opentelemtry-operator-and-collectors.md

Co-authored-by: Johannes Scheerer <johannes.scheerer@sap.com>

nickytd committed:
Update docs/proposals/34-observability2.0-opentelemtry-operator-and-collectors.md

Co-authored-by: Johannes Scheerer <johannes.scheerer@sap.com>

nickytd committed:
gep-34: clarify the explicit scope in the summary

vpnachev commented on docs/proposals/34-observability2.0-opentelemtry-operator-and-collectors.md:
```diff
@@ -0,0 +1,153 @@
+  - **Fluent-bit Integration:** Configure Fluent-bit to forward logs to the collector’s `loki` receiver, maintaining compatibility with the Vali-based setup from *GEP-19*.
+  - **Vali Integration:** Use the `loki` exporter to output the shoot logs to the existing Vali backend.
+  - **Valitail Replacement:** Replace the current `valitail` system service on shoot nodes with the OpenTelemetry collector, ensuring it collects the same types of logs as before. This will be done by deploying a systemd service for the collector on each shoot node, in the same way as the current `valitail` service.
+  - **Event Logger Replacement:** In a later stage the current `event logger` shall be replaced with the `k8s-event` receiver in  the otel-collector shoot instance
```
A link to the `k8s-event` receiver would be helpful here, is it https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/k8seventsreceiver?

rrhubenov replied:
Yes, thank you

vpnachev commented on docs/proposals/34-observability2.0-opentelemtry-operator-and-collectors.md:
```diff
@@ -0,0 +1,153 @@
+Restart=always
+RestartSec=5
+EnvironmentFile=/etc/environment
+ExecStartPre=/bin/sh -c "systemctl set-environment HOSTNAME=$(hostname | tr [:upper:] [:lower:])"
```
Dealing with the hostname will need to be done with some care, it is handled differently in the different infrastructures, some refs:
- https://github.com/gardener/gardener-extension-provider-openstack/blob/bbe65c6463add13a7f1bdbfd034e9970576ff2e5/pkg/webhook/controlplane/ensurer.go#L193
- https://github.com/gardener/gardener-extension-provider-gcp/blob/24467275e0bf2e6f8bc6713b5ba26339464094d0/pkg/webhook/controlplane/ensurer.go#L360
- https://github.com/gardener/gardener-extension-provider-aws/blob/5f900092991436b748ac8fc805adb3440d93ac54/pkg/webhook/controlplane/ensurer.go#L389

nickytd replied:
Thanks for the proposal. It makes a lot of sense. Unifying the hostname is important in regard to showing the same observability signal field/attribute in logs. That is event more important to be unified with the observability logs from the kubelet service or any other node specific service that emits logs or metrics. Synergies are mandatory in this case. That being said we will check implementation details but the way kubelets fetch the hostnames shall be utilized also by the Collector as well, ensuring consistent field/attribute values. (That is not the case in current implementation with `valitail`)

rrhubenov replied:
Since this is an implementation detail, we can discuss this at more length when the corresponding PRs are opened.
WDYT?

vpnachev replied:
Agree. You can even remove the `ExecStartPre` line from the proposal.

vpnachev reviewed (commented):
Thanks!
Overall looks good to me

nickytd committed:
Update docs/proposals/34-observability2.0-opentelemtry-operator-and-collectors.md

Co-authored-by: Johannes Scheerer <johannes.scheerer@sap.com>

rrhubenov committed:
Fix indentation & remove docker.service from _SYSTEMD_UNIT variants

rrhubenov committed:
Address PR comments (links & missing bolded component)

rfranzke commented on docs/proposals/34-observability2.0-opentelemtry-operator-and-collectors.md:
```diff
@@ -0,0 +1,153 @@
+
+### Goals
+
+- **Deploy the OpenTelemetry Operator on Seed Clusters:** Install and configure the OpenTelemetry operator to manage otel-collector instances across shoot control plane namespaces.
```
I guess we should add a goal to also deploy it in the garden runtime cluster?

nickytd replied:
On a later phase that can achieved using the documented mechanisms for deploying components on runtime cluster as well. This is out of scope for the first iteration.

rfranzke commented on docs/proposals/34-observability2.0-opentelemtry-operator-and-collectors.md:
```diff
@@ -0,0 +1,153 @@
+### Goals
+
+- **Deploy the OpenTelemetry Operator on Seed Clusters:** Install and configure the OpenTelemetry operator to manage otel-collector instances across shoot control plane namespaces.
+- **Deploy Gardener OpenTelemetry Collector Distribution in Shoot Control Planes:** Provision a dedicated Gardener OpenTelemetry collector distribution instance via the OpenTelemetry Operator in each shoot control plane namespace to handle observability data.
```
Same for virtual garden cluster?

nickytd replied:
same

rfranzke commented on docs/proposals/34-observability2.0-opentelemtry-operator-and-collectors.md:
```diff
@@ -0,0 +1,153 @@
+
+- **Objective:** Establish centralized management of OpenTelemetry collectors across all shoot control planes.
+- **Details:** Deploy the OpenTelemetry operator (e.g., from the [OpenTelemetry Operator for Kubernetes](https://github.com/open-telemetry/opentelemetry-operator)) in the `garden` namespace of each seed cluster using `ManagedResources`, similar to the Prometheus operator deployment in [*GEP-19*](https://github.com/gardener/gardener/blob/master/docs/proposals/19-migrating-observability-stack-to-op ...
+- **Configuration:** The operator will be deployed following the `botanist` workflow in the same way as the current `prometheus` and `fluent-bit` operators. This means that during the `Seed` reconciliation flow the OpenTelemetryOperator will be deployed into the `garden` namespace.
```
Which botanist workflow? Can you share a link?

nickytd replied:
I am not aware of such link or docs.

rfranzke commented on docs/proposals/34-observability2.0-opentelemtry-operator-and-collectors.md:
```diff
@@ -0,0 +1,153 @@
+        endpoint: https://v-local--local.ingress.local.seed.local.gardener.cloud/vali/api/v1/push
+        headers:
+          # Authenticate against Vali using a Bearer Token
+          Authorization: "Bearer ${file:{{ .pathAuthToken }}}"
```
This doesn't look good - we probably don't want to put sensitive data like an auth token into an unencrypted resource.

vpnachev replied:
Technically, it should be the same as the `gardener-node-agent` or `kubelet` credentials, no?

rfranzke replied:
Valitail has has its own credentials/token as of today

ScheererJ replied:
If I understand https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/extension/bearertokenauthextension correctly it should be possible to use `auth.authenticator` and `extensions.bearertokenauth.filename`, which would allow the custom resource to stay clear of any secret, but still use the collector.
@nickytd / @rrhubenov WDYT?

rrhubenov replied:
I'd like to give additional context so that we're on the same page when discussing this config. Sorry if I don't say anything new.

The configuration that we're discussing is the configuration of the `Collector` that will be deployed to the `Shoot Control Plane` via an `OpenTelemetryCollector` Custom Resource. The operator will handle the deployment of the Collector.

The `${file: <file_name>}` placeholder will load the contents of the given filename (from the {{ .pathAuthToken}} placeholder that is filled during yaml rendering in the shoot creation flow) in the configuration during the runtime of the collector. This means that the exposed part in the custom resource is the path to the file containing the token and not the token itself. The collector will expect that it has read permissions on that file and it will load it into its runtime when parsing the configuration file we're discussing. I am not sure whether having a path to the token file in the resource would be considered a vulnerability.

I'm not sure whether the comparison with the `valitail` configuration is appropriate in this context since `valitail` is running on the shoot nodes rather than the control plane. As far as I've understood in the current mechanism of `valitail` authentication, the `gardener-node-agent` syncs a secret with a file inside the node. This file contains the `valitail` auth token. The valitail config references the file and as I've outlined in the previous paragraph, it should be the same as the config we're discussing? (excluding the secret <-> file syncing from the `gardener-node-agent`)

I think this topic is interesting and can be evolved into a more general discussion of 'how would we want the node components to authenticate with the control-plane?'. If we're concerned with exposing tokens/secrets in the shoot nodes, I'd say the issue is not contained only to this discussion around the proposed collector.

With all that being said:
- Yes, the `bearertokenauthextension` that has been linked can absolutely be used. I'd even say it would be cleaner from an implementation standpoint but I'm not sure whether it handles the concern we're discussing. Thank you for proposing it, I'll most likely migrate to it rather than specifying the header itself if we continue with this approach.
- Maybe we'd like to talk about the other config that is for the collector that will be In the shoot nodes (the one that will actually replace the `valitail` component)?

rfranzke replied:
> This means that the exposed part in the custom resource is the path to the file containing the token and not the token itself. The collector will expect that it has read permissions on that file and it will load it into its runtime when parsing the configuration file we're discussing. I am not sure whether having a path to the token file in the resource would be considered a vulnerability.

Okay, then it is fine. This isn't obvious from this line. I am not aware that you can send a header like this (`Authorization: Bearer <path-to-token-file>`). Can you please adapt this YAML example here so that it becomes more clear how this works (i.e., that only the path to the token file is referenced in the resource, not the token itself)?

rrhubenov replied:
Of course. I agree that it's confusing so I'll be adapting this to the `bearertokenauthextension` that Johannes proposed. I'd first like to test it out so that I'm sure that it works as expected.

rrhubenov replied:
I can confirm that it works as expected. Will update the GEP

nickytd committed:
Align terms and capatilizations according the project recomendations.

Glossary
- https://opentelemetry.io/docs/concepts/glossary/
- https://opentelemetry.io/docs/concepts/glossary/#opentelemetry
- https://opentelemetry.io/docs/concepts/glossary/#collector

Signed-off-by: Niki Dokovski <nikolai.dokovski@sap.com>

rrhubenov committed:
Add bearertokenauth to collector in control-plane

ScheererJ commented on docs/proposals/34-observability2.0-opentelemtry-operator-and-collectors.md:
```diff
@@ -0,0 +1,160 @@
+
+## Summary
+
+This proposal introduces the OpenTelemetry Operator and deploys OpenTelemetry Collectors within the shoot control plane namespaces of Gardener-managed clusters. Building on the foundation laid by [GEP-19](https://github.com/gardener/gardener/blob/master/docs/proposals/19-migrating-observability-stack-to-operators.md) (migration to Prometheus and Fluent-bit operators) and the [Observability 2.0](h ...
```
Please use relative links when possible, i.e.

```suggestion
This proposal introduces the OpenTelemetry Operator and deploys OpenTelemetry Collectors within the shoot control plane namespaces of Gardener-managed clusters. Building on the foundation laid by [GEP-19](./19-migrating-observability-stack-to-operators.md) (migration to Prometheus and Fluent-bit operators) and the [Observability 2.0](https://github.com/gardener/logging/blob/master/docs/observability-2.0/Observability%202.0.md) vision, this GEP advances Gardener's observability stack by adopting OpenTelemetry standards. The initiative aims to layout the foundation for building a more cohesive and interoperable observability framework. This GEP takes the first step by introducing OpenTelemetry for logging, aligning with the [Observability 2.0](https://github.com/gardener/logging/blob/master/docs/observability-2.0/Observability%202.0.md) vision, with future enhancements planned to include metrics and optionally traces.
```

ScheererJ commented on docs/proposals/34-observability2.0-opentelemtry-operator-and-collectors.md:
```diff
@@ -0,0 +1,160 @@
+
+## Motivation
+
+Gardener's current observability stack, while improved by the operator-based approach in [GEP-19](https://github.com/gardener/gardener/blob/master/docs/proposals/19-migrating-observability-stack-to-operators.md), still relies on vendor specific format and protocols for collecting, processing and storing observability signals. This brings challenges in extending both components and scenarios used  ...
```
Please use relative links when possible, i.e.

```suggestion
Gardener's current observability stack, while improved by the operator-based approach in [GEP-19](./19-migrating-observability-stack-to-operators.md), still relies on vendor specific format and protocols for collecting, processing and storing observability signals. This brings challenges in extending both components and scenarios used to process these signals and enforces lock-in integration scenarios with external consumers. By introducing the OpenTelemetry Operator and Collectors, this proposal addresses these challenges, aligning Gardener with industry trends and preparing it for future observability enhancements, such as unified visualization and tracing support.
```

ScheererJ commented on docs/proposals/34-observability2.0-opentelemtry-operator-and-collectors.md:
```diff
@@ -0,0 +1,160 @@
+### 1. Deploy the OpenTelemetry Operator on Seed Clusters
+
+- **Objective:** Establish centralized management of OpenTelemetry Collectors across all shoot control planes.
+- **Details:** Deploy the OpenTelemetry Operator (e.g., from the [OpenTelemetry Operator for Kubernetes](https://github.com/open-telemetry/opentelemetry-operator)) in the `garden` namespace of each seed cluster using `ManagedResources`, similar to the Prometheus operator deployment in [*GEP-19*](https://github.com/gardener/gardener/blob/master/docs/proposals/19-migrating-observability-stack-to-op ...
```
Please use relative links when possible, i.e.

```suggestion
- **Details:** Deploy the OpenTelemetry Operator (e.g., from the [OpenTelemetry Operator for Kubernetes](https://github.com/open-telemetry/opentelemetry-operator)) in the `garden` namespace of each seed cluster using `ManagedResources`, similar to the Prometheus operator deployment in [*GEP-19*](./19-migrating-observability-stack-to-operators.md).
```

ScheererJ commented on docs/proposals/34-observability2.0-opentelemtry-operator-and-collectors.md:
```diff
@@ -0,0 +1,160 @@
+
+- **Objective:** Ensure a smooth transition by interfacing with the current observability stack.
+- **Details:**
+  - **Fluent-bit Integration:** Configure Fluent-bit to forward logs to the collector’s `loki` receiver, maintaining compatibility with the Vali-based setup from [*GEP-19*](https://github.com/gardener/gardener/blob/master/docs/proposals/19-migrating-observability-stack-to-operators.md).
```
Please use relative links when possible, i.e.

```suggestion
  - **Fluent-bit Integration:** Configure Fluent-bit to forward logs to the collector’s `loki` receiver, maintaining compatibility with the Vali-based setup from [*GEP-19*](./19-migrating-observability-stack-to-operators.md).
```

ScheererJ commented on docs/proposals/34-observability2.0-opentelemtry-operator-and-collectors.md:
```diff
@@ -0,0 +1,160 @@
+- **Description:** Deploy OpenTelemetry Collectors in each shoot reconcile directly without using an operator.
+- **Pros:** Simplifies initial setup, avoids operator overhead.
+- **Cons:** Lacks scalability and management features (e.g., automatic upgrades, CR-based configuration), which are critical in Gardener’s multi-tenant environment.
+- **Reason Rejected:** The operator aligns with [*GEP-19*](https://github.com/gardener/gardener/blob/master/docs/proposals/19-migrating-observability-stack-to-operators.md)’s precedent of using operators for observability components, ensuring consistency and scalability.
```
Please use relative links when possible, i.e.

```suggestion
- **Reason Rejected:** The operator aligns with [*GEP-19*](./19-migrating-observability-stack-to-operators.md)’s precedent of using operators for observability components, ensuring consistency and scalability.
```

ScheererJ commented on docs/proposals/34-observability2.0-opentelemtry-operator-and-collectors.md:
```diff
@@ -0,0 +1,160 @@
+- "@nickytd"
+- "@rrhubenov"
+reviewers:
+-
```
You might want to add some reviewers here. It seems at least four people looked at the GEP.

ScheererJ commented on docs/proposals/34-observability2.0-opentelemtry-operator-and-collectors.md:
```diff
@@ -0,0 +1,160 @@
+-
+---
+
+# GEP-[34]: Introducing OpenTelemetry Operator and Collectors in Shoot Control Planes
```
```suggestion
# GEP-34: Introducing OpenTelemetry Operator and Collectors in Shoot Control Planes
```

ScheererJ commented on docs/proposals/34-observability2.0-opentelemtry-operator-and-collectors.md:
```diff
@@ -0,0 +1,160 @@
+
+---
+
+This GEP establishes a critical milestone in Gardener’s journey toward *Observability 2.0*. By deploying the OpenTelemetry Operator and OpenTelemetry Collectors, Gardener gains a standardized, extensible observability framework that integrates with its existing stack while paving the way for future enhancements like tracing and OpenTelemetry native backends.
```
Please add another link to the Observability 2.0 document.

ScheererJ reviewed (commented):
Thanks for addressing the feedback. I read through the document again. It is almost finished from my point of view, but there are a few nits to be resolved.

The only real omission is the addition of the reviewers. Please take care.

rrhubenov committed:
Address feedback

rrhubenov committed:
Remove `ExecStartPre` from unit definition

rfranzke reviewed (commented):
/lgtm
